lib/SymbolGraphGen/SymbolGraph.cpp

```diff
@@ -645,29 +645,31 @@ void SymbolGraph::serialize(llvm::json::OStream &OS) {
       symbolgraphgen::serialize(*ModuleVersion, OS);
     }
 
+    // Sort symbols for deterministic output.
+    SmallVector<Symbol> Symbols(Nodes.begin(), Nodes.end());
+    llvm::sort(Symbols, [](const Symbol &LHS, const Symbol &RHS) {
+      SmallString<256> LHSUSR, RHSUSR;
+      LHS.getUSR(LHSUSR);
+      RHS.getUSR(RHSUSR);
+      return LHSUSR < RHSUSR;
+    });
     OS.attributeArray("symbols", [&](){
-      for (const auto &S: Nodes) {
+      for (const auto &S: Symbols) {
         S.serialize(OS);
       }
     });
 
-#ifndef NDEBUG
-    // FIXME (solver-based-verification-sorting): In assert builds sort the
-    // edges so we get consistent symbol graph output. This allows us to compare
-    // the string representation of the symbolgraph between the solver-based
-    // and AST-based result.
-    // This can be removed once the AST-based cursor info has been removed.
-    SmallVector<Edge> Edges(this->Edges.begin(), this->Edges.end());
-    std::sort(Edges.begin(), Edges.end(), [](const Edge &LHS, const Edge &RHS) {
+    // Sort the edges so we get consistent symbol graph output.
+    SmallVector<Edge> OrderedEdges(Edges.begin(), Edges.end());
+    llvm::sort(OrderedEdges, [](const Edge &LHS, const Edge &RHS) {
       SmallString<256> LHSTargetUSR, RHSTargetUSR;
       LHS.Target.getUSR(LHSTargetUSR);
       RHS.Target.getUSR(RHSTargetUSR);
       return LHSTargetUSR < RHSTargetUSR;
```

Comment:
I know this is code you didn't touch, but i feel like this should order on more than target USR - there can be several relationships that target the same symbol which are still going to be in a nondeterminstic ordering after this sort is applied.

Personally i would define a `<` operator for `Edge` which compares target, source, and kind, and defer to that in this sort method.

Reply:
Sounds good. Is USR good enough for `Symbol` entry?

Reply:
*Technically* no. Imported Objective-C symbols that have an automatically-converted version (e.g. a completion handler being converted into an async method) will appear in symbol graphs as two symbols with the same USR. I'm not sure how to fetch that info offhand, but being able to match those would be useful as well.

In SymbolKit we would just defer to the declaration fragments, but we don't store those outside the output JSON structure, so i'm not sure what could be used as an ordering constraint in this case.

Reply:
Maybe it is easier if just convert `DenseSet` to `llvm::SetVector`? It uses more memory but assuming the ASTWalker is deterministic, then preserving the order of insertion should make symbol graph deterministic too.

Reply:
I'm not sure if ASTWalker (SourceEntityWalker in our case) is deterministic, but if it is then that should be perfect.

Reply:
A regular ASTWalker should be, unless some function overlayed on top is breaking this, then it is a different problem to fix. I run it in a loop for a while and so far nothing is non-deterministic. Let me update to use that.